java-checks/src/main/java/org/sonar/java/checks/regex/RegexComplexityCheck.java

```diff
@@ -0,0 +1,235 @@
+    defaultValue = "" + DEFAULT_MAX)
+  private int max = DEFAULT_MAX;
+
+  @Override
+  public void checkRegex(RegexParseResult parseResult, MethodInvocationTree mit) {
+    ExpressionTree regexArgument = mit.arguments().get(0);
+    // Other than getting the flags, the parse result is not used. We find and parse the parts of
+    // the regex ourselves because we want to count the complexity of each part individually if
+    // the regex is made out of parts stored in variables.
+    FlagSet flags = parseResult.getInitialFlags();
+    for (LiteralTree[] regexPart : findRegexParts(regexArgument)) {
+      new ComplexityCalculator().visit(regexForLiterals(flags, regexPart));
+    }
+  }
+
+  List<LiteralTree[]> findRegexParts(ExpressionTree regexArgument) {
+    RegexPartFinder finder = new RegexPartFinder();
+    finder.find(regexArgument);
+    return finder.parts;
+  }
+
+  public void setMax(int max) {
+    this.max = max;
+  }
+
+  private static class RegexPartFinder {
+
+    List<LiteralTree[]> parts = new ArrayList<>();
+
+    void find(ExpressionTree expr) {
+      switch (expr.kind()) {
+        case PLUS:
+          List<LiteralTree> literals = new ArrayList<>();
+          findInStringConcatenation(expr, literals);
+          if (!literals.isEmpty()) {
+            parts.add(literals.toArray(new LiteralTree[0]));
+          }
+          break;
+        case IDENTIFIER:
+          getFinalVariableInitializer((IdentifierTree) expr).ifPresent(this::find);
+          break;
+        case PARENTHESIZED_EXPRESSION:
+          find(ExpressionUtils.skipParentheses(expr));
+          break;
+        case STRING_LITERAL:
+          parts.add(new LiteralTree[] {(LiteralTree) expr});
+          break;
+        default:
+          // Do nothing
+      }
+    }
+
+    void findInStringConcatenation(ExpressionTree expr, List<LiteralTree> literals) {
+      if (expr.is(Tree.Kind.STRING_LITERAL)) {
+        literals.add((LiteralTree) expr);
+      } else if (expr.is(Tree.Kind.PLUS)) {
+        BinaryExpressionTree binExpr = (BinaryExpressionTree) expr;
+        findInStringConcatenation(binExpr.leftOperand(), literals);
+        findInStringConcatenation(binExpr.rightOperand(), literals);
+      } else if (expr.is(Tree.Kind.PARENTHESIZED_EXPRESSION)) {
+        findInStringConcatenation(ExpressionUtils.skipParentheses(expr), literals);
+      } else {
+        find(expr);
+      }
+    }
+
+  }
+
+  private class ComplexityCalculator extends RegexBaseVisitor {
+
+    int complexity = 0;
+
+    int nesting = 1;
+
+    List<RegexIssueLocation> components = new ArrayList<>();
+
+    private RegexSyntaxElement firstComponent = null;
+
+    private void increaseComplexity(RegexSyntaxElement syntaxElement, int increment) {
+      complexity += increment;
+      if (firstComponent == null) {
+        firstComponent = syntaxElement;
+      }
+      components.add(new RegexIssueLocation(syntaxElement, "+" + increment));
```

Comment:
Looking at examples of issues from peach, I think it would be nice to add the amount caused by nesting (to be closer from cognitive complexity), such as:
```
+2 (incl 1 for nesting)
```